src/lib.rs

```diff
@@ -58,60 +56,99 @@
 //! # }
 //! ```
 
-use std::iter::Peekable;
+#![no_std]
+
+use core::iter::Peekable;
 
 /// The iterator returned by `peeking_take_while`.
 ///
 /// See the [module documentation](./index.html) for details.
 pub struct PeekingTakeWhile<'a, I, P>
-    where I: 'a + Iterator
+where
+    I: Iterator,
 {
     iter: &'a mut Peekable<I>,
     predicate: P,
+    flag: bool,
 }
 
 impl<'a, I, P> Iterator for PeekingTakeWhile<'a, I, P>
-    where I: Iterator,
-          I::Item: ::std::fmt::Debug,
-          P: FnMut(&<I as Iterator>::Item) -> bool
+where
+    I: Iterator,
+    P: FnMut(&I::Item) -> bool,
 {
-    type Item = <I as Iterator>::Item;
+    type Item = I::Item;
 
+    #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        let predicate = &mut self.predicate;
-        if self.iter.peek().map_or(false, |x| !(predicate)(x)) {
+        if self.flag {
             None
+        } else if let Some(x) = self.iter.next_if(&mut self.predicate) {
+            Some(x)
         } else {
-            self.iter.next()
+            self.flag = true;
+            None
         }
     }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let upp = if self.flag {
+            Some(0)
+        } else {
+            self.iter.size_hint().1
+        };
+        // can't know a lower bound, due to the predicate
+        (0, upp)
```

Comment:
And without the `self.flag` fusing, then this method becomes simply

```suggestion
        // Can't know a lower bound, due to the predicate.
        (0, self.iter.size_hint().1)
```